build_runner/lib/src/asset_graph/graph.dart

```diff
@@ -77,6 +80,51 @@ class AssetGraph {
   /// Gets all nodes in the graph.
   Iterable<AssetNode> get allNodes => _nodesById.values;
 
+  /// Invalidates all generated assets.
+  ///
+  /// If the build script has changed any Builder could have new behavior which
+  /// would produce a different output.
+  void invalidateBuildScript() {
+    for (var node in allNodes) {
+      if (node is GeneratedAssetNode) node.needsUpdate = true;
+    }
+  }
+
+  /// Update graph structure, invalidate outputs that may change, and return the
+  /// set of assets that need to be deleted.
+  Iterable<AssetId> updateAndInvalidate(Map<AssetId, ChangeType> updates) {
+    var deletes = new Set<AssetId>();
+    var seen = new Set<AssetId>();
+    void clearNodeAndDeps(AssetId id, ChangeType rootChangeType,
+        {AssetId parent}) {
+      if (seen.contains(id)) return;
+      seen.add(id);
+      var node = this.get(id);
+      if (node == null) return;
+
+      // Update all outputs of this asset as well.
+      for (var output in node.outputs) {
+        clearNodeAndDeps(output, rootChangeType, parent: node.id);
+      }
+
+      // For deletes, prune the graph.
+      if (parent == null && rootChangeType == ChangeType.REMOVE) {
+        remove(id);
+      }
+      if (node is GeneratedAssetNode) {
+        node.needsUpdate = true;
+        if (rootChangeType == ChangeType.REMOVE &&
+            node.primaryInput == parent) {
```

Comment:
If some other input to this action hits this codepath first, then `node.primaryInput` will not be equal to `parent`, but that doesn't mean that `node.primaryInput` wasn't removed as well.... seems like a bug.

Maybe we need a `node.primaryOutputs` so we can remove those more explicitly if `node` is removed instead of this logic?